### PR TITLE
Feature: Stuff for chocolate factory and hoppity event

### PR DESCRIPTION
**Pull request by CalMWolfs**
## Dependencies
- https://github.com/hannibal002/SkyHanni-REPO/pull/143
- #1445

## What
Add stuff for the alpha update. Info inside the chocolate inventory & Waypoints to the shared egg locations

## Changelog New Features
+ Added stuff for Chocolate Factory. - CalMWolfs
    * Show info about your Chocolate Factory.
    * Show which upgrades you can afford and which to buy.
    * Notification to click on rabbit in the inventory.
    * Notify you if you are close to having your rabbits crushed.
+ Added solver for the Egg Locator. - CalMWolfs & gravy
+ Added other stuff for Hoppity Event. - CalMWolfs
    * Display which eggs you have not gotten this SkyBlock day.
    * Waypoints for every egg on all islands.
    * Shared waypoints for found eggs.

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/config/features/misc/ChocolateFactoryConfig.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        ChocolateFactoryStatsType.ALL_TIME,
+        ChocolateFactoryStatsType.EMPTY,
+        ChocolateFactoryStatsType.PER_SECOND,
+        ChocolateFactoryStatsType.PER_MINUTE,
+        ChocolateFactoryStatsType.PER_HOUR,
+        ChocolateFactoryStatsType.PER_DAY,
+        ChocolateFactoryStatsType.EMPTY_2,
+        ChocolateFactoryStatsType.MULTIPLIER,
+        ChocolateFactoryStatsType.BARN
+    ));
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/config/features/misc/ChocolateFactoryConfig.java]: specify this list in ChocolateFactoryStatsType. Since there it won't need the prefix.

File: src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java [hunk 2]
```diff
@@ -40,6 +40,10 @@ public class MiscConfig {
     @Category(name = "Stranded", desc = "Features for the Stranded game mode.")
     public StrandedConfig stranded = new StrandedConfig();
 
+    @Expose
+    @Category(name = "Hoppity", desc = "Features for the Hoppity event.")
+    public ChocolateFactoryConfig chocolateFactory = new ChocolateFactoryConfig();
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java]: this should be in events, maybe.

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    var bestUpgrade: Int? = null
+    var bestRabbitUpgrade: String? = null
+
+    @SubscribeEvent
+    fun onInventoryOpen(event: InventoryFullyOpenedEvent) {
+        if (!LorenzUtils.inSkyBlock) return
+        if (event.inventoryName != "Chocolate Factory") return
+        inChocolateFactory = true
+
+        DelayedRun.runDelayed(50.milliseconds) {
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt]:
```suggestion
        DelayedRun.runNextTick {
```

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+        bestUpgrade = null
+        upgradeableSlots.clear()
+        var bestAffordableUpgradeRatio = Double.MAX_VALUE
+        var bestPossibleUpgradeRatio = Double.MAX_VALUE
+
+        for ((slotIndex, item) in inventory) {
+            if (config.rabbitWarning && clickMeRabbitPattern.matches(item.name)) {
+                SoundUtils.playBeepSound()
+            }
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt]: maybe also highlight the slot. (Not at this place but you know how to do it)

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+            val upgradeCost = item.getLore().getUpgradeCost() ?: continue
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt]: save the lore in val to cache it instead of calling it twice.

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+        productionInfoIndex = data.productionInfoIndex
+        prestigeIndex = data.prestigeIndex
+        milestoneIndex = data.milestoneIndex
+        maxRabbits = data.maxRabbits
+    }
+
+    @SubscribeEvent
+    fun onSecondPassed(event: SecondPassedEvent) {
+        HoppityEggType.checkClaimed()
+    }
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt]: why is this here? Maybe move it to HoppityEggManager.

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+    private fun List<String>.getUpgradeCost(): Long? {
+        val nextLine = this.nextAfter({ it == "§7Cost" }) ?: return null
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt]: "§7Cost" should be a pattern

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryBarnManager.kt [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        "rabbit.duplicate",
+        "§7§lDUPLICATE RABBIT! §6\\+[\\d,]+ Chocolate"
+    )
+
+    var barnFull = false
+    private var lastBarnFullWarning = SimpleTimeMark.farPast()
+
+    @SubscribeEvent
+    fun onChat(event: LorenzChatEvent) {
+        if (!LorenzUtils.inSkyBlock) return
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryBarnManager.kt]: this runs Always, even when the main enable is off

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+    var chocolateMultiplier = 1.0
+
+    val upgradeableSlots: MutableSet<Int> = mutableSetOf()
+    var bestUpgrade: Int? = null
+    var bestRabbitUpgrade: String? = null
+
+    @SubscribeEvent
+    fun onInventoryOpen(event: InventoryFullyOpenedEvent) {
+        if (!LorenzUtils.inSkyBlock) return
+        if (event.inventoryName != "Chocolate Factory") return
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt]: pattern

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+    var chocolateThisPrestige = 0L
+    var chocolateMultiplier = 1.0
+
+    val upgradeableSlots: MutableSet<Int> = mutableSetOf()
+    var bestUpgrade: Int? = null
+    var bestRabbitUpgrade: String? = null
+
+    @SubscribeEvent
+    fun onInventoryOpen(event: InventoryFullyOpenedEvent) {
+        if (!LorenzUtils.inSkyBlock) return
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt]: also this runs always even when the main toggle is disabled. (When included here it will also disable most other events from the feature)

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggType.kt [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+    fun isClaimed() = claimed
+    fun formattedName() = "$mealColour$mealName"
+
+    companion object {
+        fun allFound() = entries.forEach { it.markClaimed() }
+
+        fun getMealByName(mealName: String) = entries.find { it.mealName == mealName }
+
+        fun checkClaimed() {
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggType.kt]: add a the feature toggle here  (or in the callers)

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryInventory.kt [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+        ".*\\s(?<tier>[IVXLC]+)"
+    )
+    private val unclaimedRewardsPattern by ChocolateFactoryApi.patternGroup.pattern(
+        "unclaimedrewards",
+        "§7§aYou have \\d+ unclaimed rewards?!"
+    )
+
+    @SubscribeEvent
+    fun onRenderItemOverlayPost(event: GuiRenderItemEvent.RenderOverlayEvent.GuiRenderItemPost) {
+        if (!ChocolateFactoryApi.isEnabled()) return
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryInventory.kt]: isEnabled is not needed here since the inChocolateFactory should cover it. (When suggestion is implemented in ChocolateFacoryApi at InventoryFullyOpenedEvent)

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryInventory.kt [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+        val item = event.stack ?: return
+        val itemName = item.name
+        if (itemName != ChocolateFactoryApi.bestRabbitUpgrade) return
+
+        event.drawSlotText(event.x + 18, event.y, "§6✦", .8f)
+    }
+
+    @SubscribeEvent
+    fun onBackgroundDrawn(event: GuiContainerEvent.BackgroundDrawnEvent) {
+        if (!ChocolateFactoryApi.isEnabled()) return
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryInventory.kt]: isEnabled is not needed here since the inChocolateFactory should cover it. (When suggestion is implemented in ChocolateFacoryApi at InventoryFullyOpenedEvent)
x2

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryInventory.kt [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+            }
+            if (slot.slotIndex == ChocolateFactoryApi.barnIndex && ChocolateFactoryBarnManager.barnFull) {
+                slot highlight LorenzColor.RED
+            }
+        }
+    }
+
+    @SubscribeEvent
+    fun onRenderItemTip(event: RenderInventoryItemTipEvent) {
+        if (!ChocolateFactoryApi.isEnabled()) return
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryInventory.kt]: isEnabled is not needed here since the inChocolateFactory should cover it. (When suggestion is implemented in ChocolateFacoryApi at InventoryFullyOpenedEvent)
x3

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryInventory.kt [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+        if (slotNumber == ChocolateFactoryApi.milestoneIndex) {
+            item.getLore().matchFirst(unclaimedRewardsPattern) {
+                event.stackTip = "§c!!!"
+            }
+        }
+    }
+
+    @SubscribeEvent
+    fun onSlotClick(event: GuiContainerEvent.SlotClickEvent) {
+        if (!LorenzUtils.inSkyBlock) return
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryInventory.kt]: inSkyblock is not needed here since the inChocolateFactory should cover it. (When suggestion is implemented in ChocolateFacoryApi at InventoryFullyOpenedEvent)
x4

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryStats.kt [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+object ChocolateFactoryStats {
+
+    private val config get() = ChocolateFactoryApi.config
+
+    private var displayList = listOf<String>()
+
+    @SubscribeEvent
+    fun onBackgroundDraw(event: GuiRenderEvent.ChestGuiOverlayRenderEvent) {
+        if (!ChocolateFactoryApi.isEnabled()) return
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryStats.kt]: isEnabled is not needed here since the inChocolateFactory should cover it. (When suggestion is implemented in ChocolateFacoryApi at InventoryFullyOpenedEvent)
x5

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryStats.kt [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+        newList.add("§ePer Minute: §7${perMinute.addSeparators()}")
+        newList.add("§ePer Hour: §7${perHour.addSeparators()}")
+        newList.add("§ePer Day: §7${perDay.addSeparators()}")
+
+        newList.add("§eChocolate Multiplier: §7${ChocolateFactoryApi.chocolateMultiplier}")
+        newList.add("§eBarn: §7${ChocolateFactoryBarnManager.barnStatus()}")
+
+        newList.add("")
+        newList.add("")
+        newList.add("")
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryStats.kt]: So, I don't like that yo take the list and then remap it. Better way is to add the static text to the enum and the dynamic one as () -> String, since the enum can call any of them. And the performance of doing the perDay, perHour, perMinute is nothing compared to remapping the list. Then the remapping is redundant since you can simply call displayList = config.statsDisplayList.map{it.print()} (print is only a example name here)

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsLocations.kt [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+            for ((index, eggLocation) in possibleEggLocations.withIndex()) {
+                val eggLabel = "§aGuess #${index + 1}"
+                event.drawWaypointFilled(
+                    eggLocation,
+                    LorenzColor.GREEN.toColor(),
+                    seeThroughBlocks = true,
+                )
+                event.drawDynamicText(eggLocation.add(y = 1), eggLabel, 1.5)
+            }
+            return
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsLocations.kt]: this return here correct? Since the shared egg location won't work then.

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsLocations.kt [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+        val boundingBox = AxisAlignedBB(
+            firstPos.x, firstPos.y, firstPos.z,
+            secondPos.x, secondPos.y, secondPos.z
+        ).expand(5.0).clampTo(worldBoundingBox)
+
+        possibleEggLocations.clear()
+
+        // todo allow more leeway for further points later
+        // need to test while is hoppity event. Works well enough for now
+        val a = islandEggsLocations.filter {
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsLocations.kt]: a
most descriptive name xD

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsLocations.kt [hunk 20]
```diff
@@ -0,0 +1,221 @@
...
+        drawLocations = true
+    }
+
+    private fun getWorldBoundingBox(islandEggs: List<LorenzVec>): AxisAlignedBB {
+        var minX = 10000.0
+        var minY = 10000.0
+        var minZ = 10000.0
+        var maxX = -10000.0
+        var maxY = -10000.0
+        var maxZ = -10000.0
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsLocations.kt]: Double.positiveInfinit and Double.negativInfinit.
Works for any value. Since this way you lock it to 10000 (which is reasonable) but it could be bigger.

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsLocations.kt [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+            if (eggLocation.x > maxX) maxX = eggLocation.x
+            if (eggLocation.y > maxY) maxY = eggLocation.y
+            if (eggLocation.z > maxZ) maxZ = eggLocation.z
+        }
+        return AxisAlignedBB(minX, minY, minZ, maxX, maxY, maxZ).expand(10.0)
+    }
+
+    fun getCurrentIslandEggLocations(): List<LorenzVec>? {
+        return eggLocations[LorenzUtils.skyBlockIsland]
+    }
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsLocations.kt]:
```suggestion
    fun getCurrentIslandEggLocations(): List<LorenzVec>? =
        eggLocations[LorenzUtils.skyBlockIsland]
```

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsLocations.kt [hunk 22]
```diff
@@ -0,0 +1,221 @@
...
+
+    fun isValidEggLocation(location: LorenzVec): Boolean {
+        return getCurrentIslandEggLocations()?.any { it.distance(location) < 5.0 } ?: false
+    }
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsLocations.kt]:
```suggestion
    fun isValidEggLocation(location: LorenzVec): Boolean =
        getCurrentIslandEggLocations()?.any { it.distance(location) < 5.0 } ?: false
```

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsManager.kt [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+    private var lastMeal: HoppityEggType? = null
+
+    @SubscribeEvent
+    fun onWorldChange(event: LorenzWorldChangeEvent) {
+        lastMeal = null
+    }
+
+    @SubscribeEvent
+    fun onChat(event: LorenzChatEvent) {
+        if (!LorenzUtils.inSkyBlock) return
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsManager.kt]: misses an enable

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsManager.kt [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+    fun shareWaypointPrompt() {
+        val currentLocation = LocationUtils.playerLocation()
+        val meal = lastMeal ?: return
+        lastMeal = null
+
+        DelayedRun.runDelayed(1.seconds) {
+            ChatUtils.clickableChat(
+                "Click here to share the location of this chocolate egg with the server!",
+                onClick = { HoppityEggsShared.shareNearbyEggLocation(currentLocation, meal) },
+                SimpleTimeMark.now() + 30.seconds
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsManager.kt]:
```suggestion
                30.seconds.fromNow()
```

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsShared.kt [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+        "egg.shared",
+        ".*\\[SkyHanni] (?<meal>\\w+) Chocolate Egg located at x: (?<x>-?\\d+), y: (?<y>-?\\d+), z: (?<z>-?\\d+)"
+    )
+
+    @SubscribeEvent
+    fun onChat(event: LorenzChatEvent) {
+        if (!isEnabled()) return
+
+        sharedEggPattern.matchMatcher(event.message.removeColor()) {
+            println()
```
Comment (Thunderblade73) [inline src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsShared.kt]: no, we don't do that here.

**Review by Thunderblade73 — CHANGES_REQUESTED** [review]
Looked at the whole code. Logic looks good, Major thing you missed is a lot of enables.

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsManager.kt (hunk 23 above)
Comment (CalMWolfs) [reply src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsManager.kt]: there is no enable here as i need this data for if they turn it back on later

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt (hunk 9 above)
Comment (CalMWolfs) [reply src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryApi.kt]: I think the inventory name is fine to not be a pattern

File: src/main/java/at/hannibal2/skyhanni/config/features/misc/ChocolateFactoryConfig.java (hunk 1 above)
Comment (CalMWolfs) [reply src/main/java/at/hannibal2/skyhanni/config/features/misc/ChocolateFactoryConfig.java]: This is like what most of the config is like and is fine imo

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsLocations.kt (hunk 18 above)
Comment (CalMWolfs) [reply src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsLocations.kt]: yep, if its found with solver it should not render the shared one

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsLocations.kt (hunk 20 above)
Comment (CalMWolfs) [reply src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/HoppityEggsLocations.kt]: i dont think hypixel is making islands bigger than 10k x 10k anytime soon

File: src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java (hunk 2 above)
Comment (CalMWolfs) [reply src/main/java/at/hannibal2/skyhanni/config/features/misc/MiscConfig.java]: yeah i guess this makes sense

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryBarnManager.kt (hunk 8 above)
Comment (CalMWolfs) [reply src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryBarnManager.kt]: adressed in HoppityEggsManager

File: src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryStats.kt (hunk 17 above)
Comment (CalMWolfs) [reply src/main/java/at/hannibal2/skyhanni/features/misc/chocolatefactory/ChocolateFactoryStats.kt]: I do agree with you but I think that can be a change for a later time as we do this in a few places rn and we can change all at the same time

**Review by superhize — APPROVED** [review]
Everything looks fine and is working for me.

File: src/main/java/at/hannibal2/skyhanni/features/event/chocolatefactory/ChocolateFactoryApi.kt [hunk 26]
```diff
@@ -0,0 +1,240 @@
...
+import at.hannibal2.skyhanni.utils.StringUtils.matchFirst
+import at.hannibal2.skyhanni.utils.StringUtils.matchMatcher
+import at.hannibal2.skyhanni.utils.StringUtils.matches
+import at.hannibal2.skyhanni.utils.StringUtils.removeColor
+import at.hannibal2.skyhanni.utils.UtilsPatterns
+import at.hannibal2.skyhanni.utils.repopatterns.RepoPattern
+import net.minecraft.item.ItemStack
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+
+object ChocolateFactoryApi {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/event/chocolatefactory/ChocolateFactoryApi.kt]: API

**Review by hannibal002 — CHANGES_REQUESTED** [review]

File: src/main/java/at/hannibal2/skyhanni/config/features/event/ChocolateFactoryConfig.java [hunk 27]
```diff
@@ -0,0 +1,90 @@
...
+import io.github.notenoughupdates.moulconfig.annotations.ConfigOption;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class ChocolateFactoryConfig {
+
+    @Expose
+    @ConfigOption(name = "Hoppity Egg", desc = "")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/event/ChocolateFactoryConfig.java]: egg or eggs? please consistency

File: src/main/java/at/hannibal2/skyhanni/config/features/event/ChocolateFactoryConfig.java [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+    @Expose
+    @ConfigOption(name = "Chocolate Factory Stats", desc = "Show general info about your chocolate factory.")
+    @ConfigEditorBoolean
+    @FeatureToggle
+    public boolean statsDisplay = true;
+
+    @Expose
+    @ConfigOption(
+        name = "Stats List",
+        desc = "Drag text to change what displays and which order it displays in."
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/event/ChocolateFactoryConfig.java]: what is "it" in this context? sounds a bit weird.

File: src/main/java/at/hannibal2/skyhanni/features/event/chocolatefactory/ChocolateFactoryStats.kt [hunk 29]
```diff
@@ -0,0 +1,83 @@
...
+    private fun formatList(list: MutableList<String>): List<String> {
+        val newList = mutableListOf<String>()
+        for (index in config.statsDisplayList) {
+            newList.add(list[index.ordinal])
+        }
+
+        return newList
+    }
+
+    enum class ChocolateFactoryStatsType(val display: String) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/event/chocolatefactory/ChocolateFactoryStats.kt]: maybe rename to ChocolateFactoryStat

Comment (hannibal002) [reply src/main/java/at/hannibal2/skyhanni/features/event/chocolatefactory/ChocolateFactoryStats.kt]: `val` can go here.
we dont use `display` outside this class, and we dont want to ether (its an example text after all)

**Review by hannibal002 — CHANGES_REQUESTED** [review]
another small code cleanup review

**Review by hannibal002 — APPROVED** [review]
